src/shared/utils/ast/visitors.ts

```diff
@@ -0,0 +1,373 @@
+        halt,
+        whenFound,
+        acc
+      )
+    case 'FunctionExpression':
+      return recurseWithoutSharingSiblingScope(
+        [node.body],
+        halt,
+        whenFound,
+        acc
+      )
+    case 'CallExpression':
+      if (!node.arguments) {
+        return acc
+      }
+      // technically the arguments are expressions which can be evaled
+      return recurseWithoutSharingSiblingScope(
+        node.arguments,
+        halt,
+        whenFound,
+        acc
+      )
+    case 'ObjectExpression':
+      return recurseWithoutSharingSiblingScope(
+        node.properties,
+        halt,
+        whenFound,
+        acc
+      )
+    case 'MemberExpression':
+      return recurseWithoutSharingSiblingScope(
+        [node.object],
+        halt,
+        whenFound,
+        acc
+      )
+    case 'Property':
+      // technically the value is an expr which can be evaled
+      return recurseWithoutSharingSiblingScope(
+        [node.value],
+        halt,
+        whenFound,
+        acc
+      )
+    case 'IndexExpression':
+      // technically the index is an expr which can be evaled
+      return recurseWithoutSharingSiblingScope(
+        [node.array, node.index],
+        halt,
+        whenFound,
+        acc
+      )
+    case 'PipeExpression':
+      // pipe needs to share range set earlier in the pipe
+      // nested expressions (not range) will still not be shared
+      return recurseWITHSharingSiblingScope(
+        // order is important. reference AST upstream of the transformation, first.
+        [node.argument, node.call],
+        halt,
+        whenFound,
+        acc
+      )
+    case 'StringExpression':
+      return recurseWithoutSharingSiblingScope(node.parts, halt, whenFound, acc)
+    case 'InterpolatedPart':
+    case 'ParenExpression':
+      return recurseWithoutSharingSiblingScope(
+        [node.expression],
+        halt,
+        whenFound,
+        acc
+      )
+
+    // Can set Identifier
+    case 'ImportDeclaration':
+      if (node.as) {
+        return {...acc, scope: {...acc.scope, [node.as.name]: node.path}}
+      }
+      return recurseWithoutSharingSiblingScope(
+        [node.path],
+        halt,
+        whenFound,
+        acc
+      )
+
+    // Leaf in terms of setting scope, but may have the halt node (identifier)
+    case 'PackageClause':
+      return recurseWithoutSharingSiblingScope(
+        [node.name],
+        halt,
+        whenFound,
+        acc
+      )
+    case 'BuiltinStatement':
+      return recurseWithoutSharingSiblingScope([node.id], halt, whenFound, acc)
+    // Leaf
+    case 'BadStatement':
+    case 'Identifier':
+    case 'BooleanLiteral':
+    case 'DateTimeLiteral':
+    case 'DurationLiteral':
+    case 'FloatLiteral':
+    case 'PipeLiteral':
+    case 'RegexpLiteral':
+    case 'StringLiteral':
+    case 'UnsignedIntegerLiteral':
+      return acc
+    default:
+      return acc
+    // return assertUnreachable(node.type as never)
```

Comment:
This line can be uncommented during dev.

It forces typescript to make sure that the switch statement is exhausted. Meaning, that all possible use cases (based on all types that typescript knows about) are handled.